### PR TITLE
Add ROS 2 & Kubernetes use case [13412]

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
Signed-off-by: Juan López Fernández <juanlopez@eprosima.com>

### CODE REVIEW

File: docs/rst/use_cases/ros_cloud.rst [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+.. _ros_cloud:
+
+####################
+ROS 2 and Kubernetes
+####################
+
+Apart from plain LAN-to-LAN communication, Cloud environments such as container-oriented platforms have also been
+present throughout the |ddsrouter| design phase. In this walk-through example, we will set up both a *Kubernetes*
+(|k8s|) network and local environment in order to establish communication between a pair of ROS nodes, one sending
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:11]:
```suggestion
(|k8s|) network and a local environment in order to establish communication between a pair of ROS nodes, one sending
```

File: docs/rst/use_cases/ros_cloud.rst [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+participant, even when behaving as client in the participant discovery process. Make sure that the given port is
+reachable from outside this local network by properly configuring port forwarding in your Internet router device.
+The connection address points to the remote WAN participant deployed in the |k8s| cluster. For further details on how to
+configure WAN communication, please have a look at :ref:`WAN Configuration <user_manual_wan_configuration>`.
+
+To launch the local router, execute:
+
+.. code-block:: bash
+
+    ddsrouter -c local-ddsrouter.yaml
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:52]: NIT:
In general, try to use the long argument name (unless you have named it before in the page)
```suggestion
    ddsrouter --config-path local-ddsrouter.yaml
```

File: docs/rst/use_cases/ros_cloud.rst [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+.. note::
+
+    An `Ingress <https://kubernetes.io/docs/concepts/services-networking/ingress/>`_ needs to be configured for the
+    LoadBalancer service to make it externally-reachable. In this example we consider the assigned public IP address to
+    be ``2.2.2.2``.
+
+The configuration file used for the cloud router will be provided by setting up a `ConfigMap <https://kubernetes.io/docs/concepts/configuration/configmap/>`_:
+
+.. literalinclude:: ../../resources/use_cases/ros_cloud/ConfigMap.yaml
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:103]: NIT: Is there a way for the text after `|-` not to change color? Is too light

File: docs/rst/use_cases/ros_cloud.rst [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    :language: yaml
+
+Following is represented the overall configuration of our |k8s| cluster:
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:106]: Doubt: is this correct in English? And the subject?

File: docs/rst/use_cases/ros_cloud.rst [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+Now, assuming the name of the built image is ``ros2-demo-nodes:galactic``, the listener pod can be deployed by providing
+the following configuration:
+
+.. literalinclude:: ../../resources/use_cases/ros_cloud/listener.yaml
+    :language: yaml
+
+
+Once all these components are up and running, communication should have been established between talker and listener
+nodes. Feel free to interchange the locations of the ROS nodes by slightly modifying the provided configuration files,
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:143]:
```suggestion
nodes. (Maybe) Depending on the k8s platform you are using, you may have to refresh in order to see the logs of the ``stdout`` of your ``listener``.

Feel free to interchange the locations of the ROS nodes by slightly modifying the provided configuration files,
```
I would separate both sentences, as they are not related. This way, the last sentence is more visible.

**Review by jparisu — CHANGES_REQUESTED** [review]

File: docs/rst/use_cases/ros_cloud.rst (hunk 3 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:103]: Not that I know...

File: docs/rst/use_cases/ros_cloud.rst (hunk 1 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:11]: :+1:

File: docs/rst/use_cases/ros_cloud.rst (hunk 2 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:52]: :+1:

File: docs/rst/use_cases/ros_cloud.rst [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+
+
+Kubernetes setup
+================
+Two different deployments will be used for this example, each in a different |k8s| pod. The |ddsrouter| cloud instance
+(cloud router) consists of two participants; a WAN participant that receives the messages coming from our LAN through
+the aforementioned UDP communication channel, and a
+:ref:`Local Discovery Server <user_manual_participants_local_discovery_server>` (local DS) that propagates them to a
+ROS 2 listener node hosted in a different |k8s| pod. The choice of a Local Discovery Server instead of a Simple
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:76]: Good advice :+1:

File: docs/rst/use_cases/ros_cloud.rst (hunk 4 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:106]: Probably not, I changed it :+1:

File: docs/rst/use_cases/ros_cloud.rst (hunk 5 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst]: :+1:

File: docs/rst/use_cases/ros_cloud.rst [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    RMW_IMPLEMENTATION=rmw_fastrtps_cpp ros2 run demo_nodes_cpp talker
+
+
+
+Kubernetes setup
+================
+Two different deployments will be used for this example, each in a different |k8s| pod. The |ddsrouter| cloud instance
+(cloud router) consists of two participants:
+
+* A WAN participant that receives the messages coming from our LAN through the aforementioned UDP communication channel.
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:75]: Add a link to WAN Participant (even if it already exist before)
If not, it does not look fine with the link below.

File: docs/rst/use_cases/ros_cloud.rst [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+Now, assuming the name of the built image is ``ros2-demo-nodes:galactic``, the listener pod can be deployed by providing
+the following configuration:
+
+.. literalinclude:: ../../resources/use_cases/ros_cloud/listener.yaml
+    :language: yaml
+
+
+Once all these components are up and running, communication should have been established between talker and listener
+nodes, so that messages finally manage to reach the listener pod and get printed in its ``STDOUT``.
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:145]: Separate in two sentences.

File: docs/rst/use_cases/ros_cloud.rst [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+===========
+The local instance of |ddsrouter| (local router) only requires to have a
+:ref:`Simple Participant <user_manual_participants_simple>`, and a :ref:`WAN Participant <user_manual_participants_wan>`
+that will play the client role in the discovery process of remote participants
+(see :term:`Discovery Server discovery mechanism <Discovery Server>`).
+
+After having acknowledged each other’s existence through `Simple DDS discovery mechanism <https://fast-dds.docs.eprosima.com/en/latest/fastdds/discovery/simple.html>`_
+(multicast communication), the local participant will start receiving messages published by the ROS 2 talker node, and
+will then forward them to the WAN participant. Following, these messages will be sent to another participant hosted on a
+|k8s| cluster to which it connects via WAN communication over UDP/IP.
```
Comment (jparisu) [inline docs/rst/use_cases/ros_cloud.rst:27]: Add TCP comment and link.

**Review by jparisu — CHANGES_REQUESTED** [review]

File: docs/rst/use_cases/ros_cloud.rst (hunk 7 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:75]: :+1:

File: docs/rst/use_cases/ros_cloud.rst (hunk 9 above)
Comment (juanlofer-eprosima) [reply docs/rst/use_cases/ros_cloud.rst:27]: :+1:

**Review by jparisu — APPROVED** [review]
LGTM

**Pull request merged**